Add support for Bedrock Mistral models

ti3x opened the pull request:
Adding support for Mistral instruct models on AWS Bedrock:
Docs: https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-mistral.html
News: https://www.aboutamazon.com/news/aws/mistral-ai-amazon-bedrock

Test ran successfully local:
I don't think it was suspended, when I use the original code to set 
```python
#            aws_region_name=aws_region_name,
#             aws_access_key_id=aws_access_key_id,
#             aws_secret_access_key=aws_secret_access_key,
```
in completion() call, it will complain about security token. If I remove those and install boto3, everything works fine. I don't know enough about how the auth piece is set up to debug it, but it doesn't seem like an account suspension issue. 

```bash
===================================================== test session starts =====================================================
platform darwin -- Python 3.11.5, pytest-7.4.4, pluggy-1.4.0
rootdir: /Users/timx/Develop/oos/litellm/litellm/tests
plugins: anyio-3.7.1
collected 1 item                                                                                                              

test_bedrock_completion.py .                                                                                            [100%]

=============================================== 1 passed, 19 warnings in 4.71s ================================================
```

ti3x committed:
add support for bedrock mistral models

ti3x committed:
skip test but it did work locally

ti3x committed:
Add AmazonMistralConfig

ishaan-jaff commented on litellm/tests/test_bedrock_completion.py:
```diff
@@ -1,33 +1,33 @@
 # @pytest.mark.skip(reason="AWS Suspended Account")
-# import sys, os
```
My team's (litellm) account on AWS has been suspended which is why this test is commented out. Until we get access to our AWS account again

ti3x replied:
@ishaan-jaff no worries, commented them out now. I'd appreciate a quick merge by next Monday if possible since my team is eager to test out all these new Mistral models next week. Open to make any changes required though.

ishaan-jaff replied:
These tests will be re-added as soon as AWS give us access to our own account

ishaan-jaff replied:
@ti3x yes, will have this ready for by Monday

ishaan-jaff commented:
Thanks for the PR @ti3x

ti3x committed:
add assertion for test

ti3x committed:
comment out tests

krrishdholakia commented:
Hey @ti3x can you clarify what you had to comment out to make this work? (line in code would help) 

Happy to hop on a call and live debug this - https://meet.google.com/yuy-prpo-cjj

ishaan-jaff commented:
@ti3x - can you send a screenshot of this working on your side so we can merge it ?

ti3x committed:
update docs for supported models

ti3x commented:
<img width="649" alt="Screenshot 2024-03-01 at 11 38 02 PM" src="https://github.com/BerriAI/litellm/assets/65418197/34b0fdf8-b54c-4578-893b-ff20ae82bb13">

output for running
```python
from pprint import pprint
messages = [{"content": "Hello, how are you?","role": "user"}]
response = litellm.completion(
    model="bedrock/mistral.mistral-7b-instruct-v0:2",
    messages=messages,
    temperature=0
)
pprint(response.model_dump())
```

ishaan-jaff commented:
One more thing, we need this info here too: https://github.com/BerriAI/litellm/blob/2f50a660bcc693909055dbbfdac3b8f87490e14e/model_prices_and_context_window.json#L1239

ti3x commented:
> Hey @ti3x can you clarify what you had to comment out to make this work? (line in code would help)
> 
> Happy to hop on a call and live debug this - https://meet.google.com/yuy-prpo-cjj

@krrishdholakia it's almost midnight here, so I just removed the aws_ optional params, originally the test had
```
        response = completion(
            model="bedrock/anthropic.claude-instant-v1",
            messages=messages,
            max_tokens=10,
            temperature=0.1,
            aws_access_key_id=aws_access_key_id,
            aws_secret_access_key=aws_secret_access_key,
            aws_region_name=aws_region_name,
        )
```

I just changed them to 

```
        response = completion(
            model="bedrock/anthropic.claude-instant-v1",
            messages=messages,
            max_tokens=10,
            temperature=0.1,
        )
```

then it would complain about boto3. I ran `poetry add boto3` and it will pass just fine.

Unfortunately I can't live debug now, it's almost midnight around here. But let me know if anything I can provide async to help.

ti3x committed:
update prices and context window

ti3x commented:
> One more thing, we need this info here too:
> 
> https://github.com/BerriAI/litellm/blob/2f50a660bcc693909055dbbfdac3b8f87490e14e/model_prices_and_context_window.json#L1239

Added the pricing info, not sure about the mode for completion vs. chat though. Seems many instruct models also have mode="chat"

ti3x committed:
follow camelcase convention

ti3x commented on litellm/llms/bedrock.py:
```diff
@@ -282,6 +282,55 @@ def get_config(cls):
+
+    def __init__(
+        self,
+        maxTokens: Optional[int] = None,
```
@ishaan-jaff do you guys have a preference for camelcase vs. underscore? pep8 should be max_tokens.

ishaan-jaff replied:
hmm I noticed we're inconsistent in this file. Lets use camel case - since most of the other configs in this file are camel case

ti3x replied:
@ishaan-jaff I tried camelCase and it will throw an error
```
litellm.exceptions.BadRequestError: BedrockException - An error occurred (ValidationException) when calling the InvokeModel operation: Malformed input request: #: extraneous key [topP] is not permitted, please reformat your input and try again.
```

The Bedrock docs show that they expect snake case for these models
https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-mistral.html#model-parameters-mistral-request-response

ishaan-jaff replied:
You need to translate the params here @ti3x 
https://github.com/BerriAI/litellm/blob/d494941dcceb27244a2434191bcc4da1c094f0ec/litellm/utils.py#L4486

ti3x replied:
@ishaan-jaff got it, will take a look at it tomorrow.

ishaan-jaff replied:
Happy to help out in any way, we'll make sure this is in prod by Monday for you

ti3x replied:
@ishaan-jaff I added the mapping code in [a4e2476](https://github.com/BerriAI/litellm/pull/2290/commits/a4e24761a037624b11d4d565c3ad588e428d7286) so is there currently no way to map other non-openai optional params like top_k?
Works with 
```python
response = litellm.completion(
    model="bedrock/mistral.mistral-7b-instruct-v0:2",
    messages=messages,
    temperature=0,
    max_tokens=10,
    topP=0.7
)
```

I notice in the doc that a few places have             if (
                temperature == 0 and model == "mistral-7b-instruct"
            ):  # this model does no support temperature == 0
                temperature = 0.0001  # close to 0
            optional_params["temperature"] = temperature

Is this also required for Bedrock mistral-7b-instruct? Where is this 0 vs. 0.0001 documented?

ishaan-jaff replied:
> temperature == 0 and model == "mistral-7b-instruct"
): # this model does no support temperature == 0
temperature = 0.0001
> Is this also required for Bedrock mistral-7b-instruct? Where is this 0 vs. 0.0001 documented?

I believe deepinfra/mistral had this bug. I'd recommend running with bedrock and seeing if you run into this bug

ishaan-jaff replied:
> is there currently no way to map other non-openai optional params like top_k?

You can map that too, but map them in bedrock.py since they are bedrock specific

ti3x replied:
@ishaan-jaff it looks good from what I can see. the top_k param is picked up by later code branches, so it's sending over correctly
```
Request Sent from LiteLLM:

                response = client.invoke_model_with_response_stream(
                    body={"prompt": "<s>[INST] Hello, how are you? [/INST]\n</s>", "max_tokens": 10, "temperature": 0, "top_p": 0.7, "top_k": 0.5},
                    modelId=mistral.mistral-7b-instruct-v0:2,
                    accept=accept,
                    contentType=contentType
                )
```

The temperature=0 has been what I am using for testing, and it did not throw any errors from Bedrock. I could not find any Mistral-specific docs about temp=0 being unsupported so it should be all good.

ti3x committed:
change to snake case 'cause of aws docs

ti3x committed:
map optional params

ti3x committed:
update comments

ishaan-jaff commented:
@ti3x will you lmk when this is ready to review again ?

ti3x commented:
> @ti3x will you lmk when this is ready to review again ?

@ishaan-jaff I think it's ready for one final round of review.

ti3x commented:
@ishaan-jaff I think there are some minor issues with what's currently in place for the prompt template Litellm, on HF for mistral-7b and AWS docs

https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-mistral.html
https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1
and current litellm output for 
```
messages = [
    {"role": "user", "content": "What is your favourite condiment?"},
    {"role": "assistant", "content": "Well, I'm quite partial to a good squeeze of fresh lemon juice. It adds just the right amount of zesty flavour to whatever I'm cooking up in the kitchen!"},
    {"role": "user", "content": "Do you have mayonnaise recipes?"}
]
```
is 
```
<s>[INST] What is your favourite condiment? [/INST]\n Well, I'm quite partial to a good squeeze of fresh lemon juice. It adds just the right amount of zesty flavour to whatever I'm cooking up in the kitchen! [INST] Do you have mayonnaise recipes? [/INST]\n</s>
```

The </s> seems misplaced

HF version:

```
"<s>[INST] What is your favourite condiment? [/INST]"
"Well, I'm quite partial to a good squeeze of fresh lemon juice. It adds just the right amount of zesty flavour to whatever I'm cooking up in the kitchen!</s> "
"[INST] Do you have mayonnaise recipes? [/INST]"
```

AWS Doc version added role information as part of the [INST], which seems even more strange. I might have to ping them to ask for more clarifications. I can create a different messages to prompt function to do the HF version.
```
 prompt = """<s>[INST]\"role\": \"user\", \"content\": \"In Bash, how do I list 
            all text files in the current directory (excluding subdirectories)
            that have been modified in the last month?\"[/INST]"""
```

It is currently not affecting the feature from working so maybe not a PR blocker for now.

ti3x committed:
Fix message template based on transformers chat_template docs

ti3x commented:
@krrishdholakia I added https://github.com/BerriAI/litellm/pull/2290/commits/446cc393f92fffe77833eba47bc32f357188a582
which will impact all the Mistral prompt templates. Could you double check on https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2? I thought about creating a different function to minimize impact, but honestly, if the prompt template is wrong, it may be better to fix all in one place. The post-update request body:

```bash
Request Sent from LiteLLM:

            response = client.invoke_model(
                body={"prompt": "<s>[INST] What is your favourite condiment? [/INST]\n Well, I'm quite partial to a good squeeze of fresh lemon juice. It adds just the right amount of zesty flavour to whatever I'm cooking up in the kitchen!</s> [INST] Do you have mayonnaise recipes? [/INST]\n", "max_tokens": 10, "temperature": 0, "top_p": 0.7, "top_k": 0.5},
                modelId=mistral.mixtral-8x7b-instruct-v0:1,
                accept=accept,
                contentType=contentType
            )
```

ti3x commented:
@ishaan-jaff any update on merging this PR?

ishaan-jaff commented on litellm/llms/bedrock.py:
```diff
@@ -729,6 +789,9 @@ def completion(
             outputText = response_body["generation"]
+        elif provider == "mistral":
+            outputText = response_body["outputs"][0]["text"]
+            model_response["finish_reason"] = response_body["outputs"][0]["stop_reason"]
```
not a blocker to merge - but did you test this with streaming @ti3x ? 

I suspect this PR need some more work for streaming support

ti3x replied:
@ishaan-jaff it seems to work on my end, see https://github.com/BerriAI/litellm/pull/2290#discussion_r1510017323

ishaan-jaff commented on litellm/llms/prompt_templates/factory.py:
```diff
@@ -110,9 +110,9 @@ def mistral_instruct_pt(messages):
+            "assistant": {"pre_message": " ", "post_message": "</s> "},
         },
-        final_prompt_value="</s>",
+        final_prompt_value="",
```
looks correct based on this: https://huggingface.co/docs/transformers/main/chat_templating

ishaan-jaff reviewed (approved):
LGTM ! but @ti3x does streaming work? I suspect more work is required for mistral streaming on bedrock

ishaan-jaff commented:
will merge this in the next release - ETA is 1 hour from now if all goes well

ishaan-jaff commented:
@ti3x this is live! can we hop on a call sometime this week ? I'd love to learn how you're using litellm and how we can improve for you. Sharing my cal here for your convenience: https://calendly.com/d/4mp-gd3-k5k/berriai-1-1-onboarding-litellm-hosted-version?month=2023-10
